# Mercedes Benz Greener Manufacturing With Machine Learning

### import libraries

In [15]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
import gc
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [16]:
# lets use pandas to load csv files we have
train = pd.read_csv(r"C:/Users/Nageswaran B/Documents/machine_learning/machine_learning_project/simplilearn/Mercedes-Benz_Greener_Manufacturing/train.csv")
print(f"total train datapoints = {len(train)}")
train.head()

total train datapoints = 4209


ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

### Here we can see there are 4209 datapoints indexing from 0 to 4208 and 378 columns/features.
### We have three types of data in the dataset:
### float64(1): Dependent feature, testing time in seconds
### int64(369): Independent Binary features
### object(8): Independent Categorical features

In [17]:
# lets use pandas to load csv files we have
test = pd.read_csv('test.csv')
print(f"total test datapoints = {len(test)}")
test.head()

total test datapoints = 4209


ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   
2   3  az  v  as  f  d  a  j  j    0  ...     0     0     0     1     0     0   
3   4  az  l   n  f  d  z  l  n    0  ...     0     0     0     1     0     0   
4   5   w  s  as  c  d  y  i  m    0  ...     1     0     0     0     0     0   

   X382  X383  X384  X385  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 377 columns]

### Here we can see there are 4209 datapoints indexing from 0 to 4208 and 377 columns/features.
### We have three types of data in the dataset:
### int64(369): Independent Binary features
### object(8): Independent Categorical features
### We can see here we have the same number of data points in the train and test dataset.
### Now Let’s start the actual EDA,

# Statistical description of ID and Dependent variables:

In [18]:
# check if there are any missing values
isNan = train.isnull().sum().any()

if isNan == True:
    print("There are NaN values in the dataset")
else:
    print("There are No NaN values in the dataset")

There are No NaN values in the dataset


#### Let’s also check for duplicate rows,

In [19]:
# check if there are any Duplicate rows
isDup = train.duplicated().sum().any()
if isDup == True:
    print("There are Duplicate rows in the dataset")
else:
    print("There are No Duplicate rows in the dataset")

There are No Duplicate rows in the dataset


#### Now let’s check if there are any duplicate values in the ‘ID’ column,

In [20]:
dup_ID = train['ID'].duplicated().sum()
print(f"Here we have {dup_ID} duplicate IDs")

Here we have 0 duplicate IDs


# Feature Engineering:

##### LabelEncoder for encoding categorical features

In [21]:
def preprocess_categorical(data, IDs):
    """
    data : pandas dataframe
    IDs: ID feature
    return: dataframe, labels
    
    This function takes the dataframe as input, 
    encodes and normalizes the 
    categorical features.
    """
    # create empty lists for collecting feature names
    cat_features = []
    Binary_features = []
    
    # Collect the categorical and binary feature names 
    for f in data.columns:
        if data[f].dtype == 'object':
            cat_features.append(f)
        elif data[f].dtype == 'int' and f != 'ID':
            Binary_features.append(f)
        
    # create categorical feature dataframe
    cat_df = data[cat_features]
    # create binary feature dataframe
    bin_df = data[Binary_features]
    bin_df.insert(0, 'ID', IDs.values)
    
    # Now encode each categorical feature
    for feature in cat_features:
        encoder = LabelEncoder()
        cat_df[feature] = encoder.fit_transform(cat_df[feature].values)
    # Create new categorical feature dataframe
    cat_df = pd.DataFrame(cat_df, columns = cat_features)
    cat_df.insert(0, 'ID', IDs.values)
    # Merge binary and categorical dataframes together
    new_data = pd.merge(cat_df, bin_df, on='ID', how='left')
    # return dataframe and labels
    if 'y' in data.columns:
        labels = data['y']
        return new_data, labels
    else:
        return new_data

# Encode the training and testing dataset:

In [22]:
X_train, y_train = preprocess_categorical(final_train, train_ID)
X_train.head()

NameError: name 'final_train' is not defined